crates/fraud_proofs/global_merkle_root/storage/src/update.rs

```diff
@@ -0,0 +1,305 @@
+            }
+            Input::MessageCoinSigned(MessageCoinSigned { nonce, .. })
+            | Input::MessageCoinPredicate(MessageCoinPredicate { nonce, .. }) => {
+                self.storage.storage_as_mut::<Messages>().remove(nonce)?;
+            }
+            // The messages below are retryable, it means that if execution failed,
+            // message is not spend.
+            Input::MessageDataSigned(MessageDataSigned { nonce, .. })
+            | Input::MessageDataPredicate(MessageDataPredicate { nonce, .. }) => {
+                // TODO(#2589): Figure out how to know the status of the execution.
+                //  We definitely can do it via providing all receipts and verifying
+                //  the script root. But maybe we have less expensive way.
+                let success_status = false;
+                if success_status {
+                    self.storage.storage_as_mut::<Messages>().remove(nonce)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn process_output(
+        &mut self,
+        tx_pointer: TxPointer,
+        utxo_id: UtxoId,
+        inputs: &[Input],
+        output: &Output,
+    ) -> anyhow::Result<()> {
+        match output {
+            Output::Coin {
+                to,
+                amount,
+                asset_id,
+            }
+            | Output::Change {
+                to,
+                amount,
+                asset_id,
+            }
+            | Output::Variable {
+                to,
+                amount,
+                asset_id,
+            } => {
+                self.insert_coin_if_it_has_amount(
+                    tx_pointer, utxo_id, *to, *amount, *asset_id,
+                )?;
+            }
+            Output::Contract(contract) => {
+                self.try_insert_latest_contract_utxo(
+                    tx_pointer, utxo_id, inputs, *contract,
+                )?;
+            }
+            Output::ContractCreated { contract_id, .. } => {
+                self.storage.storage::<ContractsLatestUtxo>().insert(
+                    contract_id,
+                    &ContractUtxoInfo::V1((utxo_id, tx_pointer).into()),
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    fn insert_coin_if_it_has_amount(
+        &mut self,
+        tx_pointer: TxPointer,
+        utxo_id: UtxoId,
+        owner: Address,
+        amount: Word,
+        asset_id: AssetId,
+    ) -> anyhow::Result<()> {
+        // Only insert a coin output if it has some amount.
+        // This is because variable or transfer outputs won't have any value
+        // if there's a revert or panic and shouldn't be added to the utxo set.
+        if amount > Word::MIN {
+            let coin = CompressedCoinV1 {
+                owner,
+                amount,
+                asset_id,
+                tx_pointer,
+            }
+            .into();
+
+            self.storage.storage::<Coins>().insert(&utxo_id, &coin)?;
```

Comment:
Not sure, so just leaving the comment here.

Should we consider using `replace` here and check if there's been a coin in the storage already?

Reply:
Hmm, I think this is fine but it wouldn't hurt to be extra conservative. There should never be a coin that we overwrite, so sure let's do it.

Reply:
Done in https://github.com/FuelLabs/fuel-core/pull/2553/commits/6aa1c2f2b436b056c6a78ee88796c933fe6972ad